pkg/contracts/sui/withdraw_and_call_ptb.go

```diff
@@ -0,0 +1,222 @@
+	"github.com/pkg/errors"
+)
+
+const (
+	// FuncWithdrawImpl is the gateway function name withdraw_impl
+	FuncWithdrawImpl = "withdraw_impl"
+
+	// ModuleConnected is the Sui connected module name
+	ModuleConnected = "connected"
+
+	// FuncOnCall is the Sui connected module function name on_call
+	FuncOnCall = "on_call"
+
+	// TypeSeparator is the separator for Sui package and module
+	TypeSeparator = "::"
+
+	// ptbWithdrawAndCallCmdCount is the number of commands in the PTB withdraw and call
+	// the three commands are: [withdraw_impl, transfer_objects, on_call]
+	ptbWithdrawAndCallCmdCount = 3
+
+	// ptbWithdrawImplInputCount is the number of inputs in the withdraw_impl command
+	// the inputs are: [gatewayObject, amount, nonce, gasBudget, withdrawCap]
+	ptbWithdrawImplInputCount = 5
+)
+
+// ptbWithdrawImplArgIndexes is the indexes of the inputs for the withdraw_impl command
+// these are the corresponding indexes for arguments: [gatewayObject, amount, nonce, gasBudget, withdrawCap]
+// the withdraw_impl command is the first command in the PTB, so the indexes will always be [0, 1, 2, 3, 4]
+var ptbWithdrawImplArgIndexes = []int{0, 1, 2, 3, 4}
+
+// WithdrawAndCallPTB represents data for a Sui withdraw and call event
+type WithdrawAndCallPTB struct {
+	PackageID string
+	Module    string
+	Function  string
+
+	Amount math.Uint
+	Nonce  uint64
+}
+
+// TokenAmount returns the amount of the withdraw and call
+func (d WithdrawAndCallPTB) TokenAmount() math.Uint {
+	return d.Amount
+}
+
+// TxNonce returns the nonce of the withdraw and call
+func (d WithdrawAndCallPTB) TxNonce() uint64 {
+	return d.Nonce
+}
+
+// ExtractInitialSharedVersion extracts the object initial shared version from the object data
+func ExtractInitialSharedVersion(objData models.SuiObjectData) (uint64, error) {
+	owner, ok := objData.Owner.(map[string]any)
+	if !ok {
+		return 0, fmt.Errorf("invalid object owner type %T", objData.Owner)
+	}
+
+	shared, ok := owner["Shared"]
+	if !ok {
+		return 0, fmt.Errorf("missing shared object")
+	}
+
+	sharedMap, ok := shared.(map[string]any)
+	if !ok {
+		return 0, fmt.Errorf("invalid shared object type %T", shared)
+	}
+
+	return extractInteger[uint64](sharedMap, "initial_shared_version")
+}
+
+// parseWithdrawAndCallPTB parses withdraw and call with PTB.
+// There is no actual event on gateway for withdraw and call, but we construct our own event to make the logic consistent.
+func (gw *Gateway) parseWithdrawAndCallPTB(
+	res models.SuiTransactionBlockResponse,
+) (event Event, content OutboundEventContent, err error) {
+	tx := res.Transaction.Data.Transaction
+
+	// the number of PTB commands should be 3
+	if len(tx.Transactions) != ptbWithdrawAndCallCmdCount {
+		return event, nil, errors.Wrapf(
+			ErrParseEvent,
+			"invalid number of commands(%d) in the PTB",
+			len(tx.Transactions),
+		)
+	}
+
+	// the number of PTB inputs should be >= 5
+	if len(tx.Inputs) < ptbWithdrawImplInputCount {
```

Comment:
Should this have equal sign?
```suggestion
	if len(tx.Inputs) != ptbWithdrawImplInputCount {
```

Reply:
The `tx.Inputs` contains inputs of all sub commands in the PTB transaction. In the E2E case, the length is `11`.